### ISSUE TITLE
pcolor LogoList cast error on clear-all in NetLogo 3D 6.0.2

### ISSUE THREAD

**Issue by gdower** [NetLogo/NetLogo#1550]
Hello,

I'm trying to convert a model from NetLogo3D 5.3.1 to 6.0.2 and I get the following Java exception when running clear-all after the model has been setup more than once:

```
error (ClassCastException)
 while observer running CLEAR-ALL
  called by Command Center

NetLogo is unable to supply you with more details about this error.  Please report the problem
 at https://github.com/NetLogo/NetLogo/issues, or to bugs@ccl.northwestern.edu, and paste the
contents of this window into your report.

java.lang.ClassCastException: org.nlogo.core.LogoList cannot be cast to java.lang.Double
 at org.nlogo.agent.Patch3D.pcolorDoubleUnchecked(Patch3D.java:243)
 at org.nlogo.agent.AgentManagement.clearPatches(AgentManagement.scala:133)
 at org.nlogo.agent.AgentManagement.clearPatches$(AgentManagement.scala:129)
 at org.nlogo.agent.World.clearPatches(World.scala:109)
 at org.nlogo.agent.AgentManagement.clearAll(AgentManagement.scala:80)
 at org.nlogo.agent.AgentManagement.clearAll$(AgentManagement.scala:78)
 at org.nlogo.agent.World3D.clearAll(World3D.scala:362)
 at org.nlogo.workspace.AbstractWorkspaceScala.clearAll(AbstractWorkspaceScala.scala:74)
 at org.nlogo.window.GUIWorkspace.clearAll(GUIWorkspace.java:240)
 at org.nlogo.prim.etc._clearall.perform(_clearall.scala:12)
 at org.nlogo.nvm.Context.stepConcurrent(Context.java:107)
 at org.nlogo.nvm.ConcurrentJob.step(ConcurrentJob.scala:65)
 at org.nlogo.job.JobThread.runPrimaryJobs(JobThread.scala:133)
 at org.nlogo.job.JobThread.$anonfun$run$1(JobThread.scala:68)
 at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
 at scala.util.control.Exception$Catch.apply(Exception.scala:224)
 at org.nlogo.api.Exceptions$.handling(Exceptions.scala:41)
 at org.nlogo.job.JobThread.run(JobThread.scala:66)

NetLogo 3D 6.0.2
main: org.nlogo.app.AppFrame
thread: JobThread
Java HotSpot(TM) 64-Bit Server VM 1.8.0_141 (Oracle Corporation; 1.8.0_141-b15)
operating system: Linux 4.13.0-32-generic (amd64 processor)
Scala version 2.12.2
JOGL: 2.3.2
OpenGL graphics: Mesa DRI Intel(R) HD Graphics P630 (Kaby Lake GT2)
OpenGL version: 3.0 Mesa 17.2.4
OpenGL vendor: Intel Open Source Technology Center
model: debug

12:32:24.652 RuntimeErrorEvent (org.nlogo.app.App$$anon$1 (org.nlogo.window.GUIWorkspace)) AWT-EventQueue-0
12:32:24.581 AddJobEvent (org.nlogo.app.common.CommandLine) AWT-EventQueue-0
12:32:24.580 OutputEvent (org.nlogo.app.common.CommandLine) AWT-EventQueue-0
12:32:24.579 CompiledEvent (org.nlogo.window.CompilerManager) AWT-EventQueue-0
12:32:24.570 CompileMoreSourceEvent (org.nlogo.app.common.CommandLine) AWT-EventQueue-0
12:32:24.415 InterfaceGlobalEvent (org.nlogo.app.interfacetab.InterfacePanel$$anon$1 (org.nlogo.window.SliderWidget)) AWT-EventQueue-0
12:32:24.415 InterfaceGlobalEvent (org.nlogo.app.interfacetab.InterfacePanel$$anon$1 (org.nlogo.window.SliderWidget)) AWT-EventQueue-0
12:32:24.415 InterfaceGlobalEvent (org.nlogo.app.interfacetab.InterfacePanel$$anon$1 (org.nlogo.window.SliderWidget)) AWT-EventQueue-0
12:32:24.415 InterfaceGlobalEvent (org.nlogo.app.interfacetab.InterfacePanel$$anon$1 (org.nlogo.window.SliderWidget)) AWT-EventQueue-0
12:32:24.415 InterfaceGlobalEvent (org.nlogo.app.interfacetab.InterfacePanel$$anon$1 (org.nlogo.window.SliderWidget)) AWT-EventQueue-0
```

From the stack trace, I assume the exception is caused by pcolor RGBA lists not being castable to double. If I reset the patches to black first and run clear-all it fixes the problem.

Here's the code that I'm using to assign patch colors which gets called during setup:

```
; observer context
to display-normal-view
  foreach inside-patches [ inside-patch -> ask inside-patch [set pcolor [255 255 255 0]] ]
  foreach submerged-patches [ submerged-patch -> ask submerged-patch [set pcolor [186 91 34 3]] ]
  foreach pwall-patches [ pwall-patch -> ask pwall-patch [set pcolor [255 255 255 1]] ]
  foreach bottom-patches [ bottom-patch -> ask bottom-patch [set pcolor [255 255 255 1]] ]
  set prev-display-mode "normal"
end
```

Color assignment works as expected during the 1st setup, but clear-all crashes any subsequent setup calls. My temporary workaround is to just reset patches to black before running clear-all in setup:

```
to setup
  ask patches [set pcolor 0]
  clear-all
  ...
  display-normal-view
  reset-ticks
end
```

Please let me know if you need any additional information.

Thanks,

Geoff

**Comment by gdower** [comment]
I checked and this bug only affects the NetLogo3D version. Attached is code that reproduces the bug:

[pcolor_cast_bug.nlogo3d.tar.gz](https://github.com/NetLogo/NetLogo/files/1704352/pcolor_cast_bug.nlogo3d.tar.gz)

**Comment by mrerrormessage** [comment]
Thanks for reporting! We've seen issues similar to this one crop up before but haven't been able to accurately reproduce.

**Comment by mrerrormessage** [comment]
This is a straighforward fix: `Patch3D.pcolorDoubleUnchecked` needs to verify [on this line](https://github.com/NetLogo/NetLogo/blob/hexy/netlogo-core/src/main/agent/Patch3D.java#L243) that the variable is a `Double` before casting it to a `Double`.

**Comment by gdower** [comment]
I was working on submitting a patch yesterday based on the [2D version of pcolorDoubleUnchecked](https://github.com/NetLogo/NetLogo/blob/hexy/netlogo-core/src/main/agent/Patch.java#L382) and it built successfully but I couldn't figure out how to run the 3D version of NetLogo from my git clone to test it. What's the sbt command for running the 3D version?

  ```
  @Override
  public void pcolorDoubleUnchecked(Double boxedColor) {
    double color = boxedColor.doubleValue();
    Object currentColor = _variables[VAR_PCOLOR3D];
    if (!(currentColor instanceof Double) || ((Double) currentColor).doubleValue() != color) {
      _variables[VAR_PCOLOR3D] = boxedColor;
      _world.patchColors()[(int) _id] = Color.getARGBbyPremodulatedColorNumber(color);
      _world.markPatchColorsDirty();
      if (color != 0.0) {
        _world.patchesAllBlack(false);
      }
    }
  }
```

**Comment by mrerrormessage** [comment]
Awesome! The command for running 3D NetLogo from sbt is `netlogo/threed` followed by `netlogo/run`. If/when you have it working, go ahead and open a Pull Request. Thanks so much!

**Comment by gdower** [comment]
The patch works. I'll see if I can figure out how to write the unit test and then I'll open a pull request. I'll probably have it done sometime over the weekend.

**Comment by mrerrormessage** [comment]
Awesome! If you would like to push it up, I would be happy to add the test for it. The test should probably go in "test/commands/RGB.txt" and look something like the following:

```
ClearPatchRGBColor_3D
  P> set pcolor [0 0 0]
  O> clear-all
  count patches with [pcolor = [0 0 0]] => 0
```

(this is a skeleton, if there are other interactions that need to happen for the condition to occur, feel free to add those!). You can run this test by running `netlogo/test:medium` in sbt.

**Comment by mrerrormessage** [comment]
Fixed in #1557
